### DISCUSSION TITLE
API定義を考える

### DISCUSSION THREAD

**Discussion by woodnx** [ARCircle/Kojirer#9] [Ideas]
フロントエンドの各端末がどんなデータがほしいのか，API定義を考える

**Comment by tenk-9** [comment]
@4416kidr フロントエンドで必要になるAPIの定義はここでしようと思っています

### API定義までの流れ
1. フロントエンド全体の構成を定義する
    - どのような端末があり，どのような役割を持たせるのか
3. それぞれの端末で，DBに対してどのような操作をしたいか考える
4. 行う操作に基づいて以下を決める
    - 必要なAPIのレスポンスモデル（JSONで例示するのがいいですかね？それともOpenAPI documentとか？）
    - エンドポイント例
4. 決めたAPIの定義に対して，issueを立てる
5. issue立てたAPIに対するやり取り（主にbackendの人たちとのやり取り）はissueのコメントで行う

また，discussionのスレッドが長くなりすぎるのも好ましくないと思っているので，
- このdiscussionでは上の流れの2番目までを行う
- 端末ごとにdiscussionを立て，欲しいAPIの詳細定義は端末単位のフォーカスとする

のでどうでしょうか．
### このようにする理由
端末からDBに対して行いたい操作は多岐にわたると思うので，
- discussionsでどのようなAPIが欲しいか議論する
- 決まったAPIについての議論はそのissueで行う

ようにすることで，議論のフォーカスを限定したいという意図があります．
1つのスレッドで複数の議題があると混乱するので，なるべくそれは避けたいです．

### フロントエンド完成までの流れ
次のような流れで開発を進めようと思っています．これは案なので，変わりうります．

1. 欲しいAPIの定義を決め，issueでbackendチームに実装を依頼する
2. APIのレスポンス形式が定まっているので，`仮のレスポンス.json`が作れて，それを用いてページの実装をする
3. APIが実装されたら，DBにテストデータ等を入れてみて，想定通りの挙動をするかテストする
4. ダメだったら修正

自分はWeb開発経験がほとんどないので，進め方等 @4416kidr さんの意見を聞きたいです

**Comment by tenk-9** [comment]
文字ベースでの意思決定はどうしても遅くなりがちな気がするので，時間がかかりそうであれば会議やるのも手です．

**Comment by woodnx** [comment]
### 新しい丼を新規作成
- どっちか迷ってる
  - Order単位で配列としてPOST
  - 丼単位でPOST
- issueに投げます

### 丼のサイズ/トッピングに応じて値段を取得
- 丼1つに対してGETメソッドで送る．
- 返ってくるデータは値段(number)．
  - { price: 500 } とかJSONで

### 指定する`調理状態`の丼を`n`件取得
- `調理状態`と，`取得件数`を指定して丼の配列を取得
- 返ってくるデータは
```
[
  {
    dons_id,
    orders_id,
    call_num,
    yasai,
    ninnniku,
    karame,
    abura,,
    toppings: [
     currey: 1,
   ],
  },
  ...
]
```
### 指定する丼の`調理状態`を変更
- 指定するdons_idの`調理状態`を変更する
- 成功したことを意味するなにか（メッセージ）を返す．

### オプションに関して，量と対応する数値の対応表を取得

オプションの量（なし，少な目，普通，マシ，マシマシ）と，DB 内でそれらの量を表す numer の対応表を得たい．

-   request
    -   GET
    -   no params
-   returns (expected)
    -   量とそれに対応する数値の map (JSON 形式)
    -   例えば次のよう
    ```json
    {
        "無し": 0,
        "少なめ": 1,
        "普通": 2,
        "マシ": 3,
        "マシマシ": 4,
    }

**Reply by 4416kidr** [reply]
# 用語の定義

- order: 注文
  - 複数の商品から構成される注文を表現する
- dons: 丼
  - 商品1つを表現する
  - 主にsize, topping, optionから構成される
- call_num: 呼び出し番号
  - 呼び出しに使うプラカードの番号
- size: 麺の多さ
  - 多、並、少など
- option: 麺の上に乗っける具材
  - ヤサイ、ニンニク、カラメ、アブラなどを含む
  - 商品の大きさ（カロリーなど）に関係するが、料金には関係ない
- topping: 具材の上に乗っける調味料
  - カレー粉, マヨネーズなどを含む
    - 牛丼でいうチーズやとろろなど
  - 料金に関係する

**Reply by Atotti** [reply]
@tenk-9 
ステータスコードという語がIssue内で出てきますが、HTTP レスポンスステータスコードと紛らわしいので使わないで欲しい気持ちがあります。

**Reply by tenk-9** [reply]
@Atotti 失礼しました．紛らわしくてすみません．
Issue内の該当単語を置き換えたので，確認お願いします（まだ紛らわしかったり，修正されていない箇所があったらすみません）．

**Comment by tenk-9** [comment]
![IMG20240314203918.jpg](https://github.com/ARCircle/Kojirer/assets/49711514/7517fc1a-8f4d-4f20-97f0-8c1951cff22a)

**Comment by tenk-9** [comment]
この後考えなければならないのは，
- 会計時の手続きを行うエンドポイント
  - 売上個数，売上金額等を返すもの
  - 詳細（どんな情報が欲しいか）は会計と相談

です．詳細は去年会計をしていた人と相談．
